### PR TITLE
Add a left margin to TimeGraph

### PR DESCRIPTION
**Pull request by pierricgimmig**
It is assumed at various places that the thread tracks are positioned at 0 horizontally. To uncover those assumptions, and for easier future layout modifications, we add a left margin to time graph layout. It is set to zero, so there is no functional change in this PR. To visualize the effect, the "Left Margin" property has been exposed to the "devmode" layout UI.

### CODE REVIEW

**Comment by florian-kuebler** [comment]
Reassigning to @reichlfl. Code-wise it looks good. But to be honest, I lost a bit of the overview of the ui layouting part. @karupayun and @reichlfl are more into that.

File: src/OrbitGl/CallstackThreadBar.cpp [hunk 1]
```diff
@@ -118,12 +118,15 @@ void CallstackThreadBar::DoUpdatePrimitives(PrimitiveAssembler& primitive_assemb
   const Color kGreyError(160, 160, 160, 255);
   ORBIT_CHECK(capture_data_ != nullptr);
 
+  uint32_t view_element_x = viewport_->WorldToScreen(GetPos())[0];
```
Comment (reichlfl) [inline src/OrbitGl/CallstackThreadBar.cpp:121]: This should not be screen pos, but world pos (so `= GetPos()[0]`). You're using this in world space later. It doesn't make a visual difference since there is still a 1:1 mapping between those spaces, but it's conceptually incorrect.

**Review by reichlfl — CHANGES_REQUESTED** [review]

File: src/OrbitGl/CaptureViewElement.cpp [hunk 2]
```diff
@@ -127,6 +127,10 @@ void CaptureViewElement::SetVisible(bool value) {
   RequestUpdate();
 }
 
+float CaptureViewElement::HorizontalClamp(float pos_x) const {
+  return std::clamp(pos_x, pos_[0], pos_[0] + width_);
```
Comment (karupayun) [inline src/OrbitGl/CaptureViewElement.cpp:131]: Interesting. I assume that you are doing this so events are not drawn outside of the area of their track.

This is not-consistent with what is happening in the right_margin (and actually won't work there because the size of the box can overflow the margin). Also a line outside of the left-most part might be visible misleading the user. In the right-margin area we are rendering an entire rectangle with background_color on top, so events in that area won't be visible since they are behind the rectangle. So:

1- Either instead of clamping, you do the same than the right-margin and draw another rectangle for the `left_margin`.

2- Aim for a proper solution that will work for both margins to later erase the right_margin.
I have some ideas about this "proper solution" but I'm also interested in @reichlfl opinion.

I thought on doing this on the "Batcher/PrimitiveAssembler" level. Instead of the current implementation, all methods to add primitives will take an additional Pos/Size parameter from the container that will limit the area where the primitives will be drawn (and possibly no draw anything at all). It's a big refactor though but it will be error-prone as we won't need to clamp everything before drawing.

What do you think? @reichlfl @pierricgimmig

**Review by karupayun — COMMENTED** [review]
Thank you very much Pierric! This PR was really needed :)

I left some comments and I will work in parallel to fix the GetPixelNumber bug

File: src/OrbitGl/CallstackThreadBar.cpp (hunk 1 above)
Comment (karupayun) [reply src/OrbitGl/CallstackThreadBar.cpp:121]: https://github.com/google/orbit/pull/4494

Comment (reichlfl) [reply src/OrbitGl/CallstackThreadBar.cpp:121]: Thanks Pablo

File: src/OrbitGl/CaptureViewElement.cpp (hunk 2 above)
Comment (reichlfl) [reply src/OrbitGl/CaptureViewElement.cpp:131]: 2 would be my preferred solution as it neatly scales to all elements, and you don't need to take care about it when implementing draw methods.

But I'm not convinced that adding it as parameters to PrimitiveAssembler calls is the right thing to do, it's *a lot* of parameters that we'd need to add all across the place, its a big refactor and takes a lot of additional mental overhead when writing code. Instead, I would add this into the Batcher state, akin to `Push/PopTranslation`. Each `CaptureViewElement` pushes its bounding rect onto the batcher before drawing, and pops it afterwards, that way the draw methods are basically unaffected. It may still be quite a bit of work to verify everything still works as expected, and it might also have significant performance impact.

So in conclusion, I think this localized fix is also fine for me. But I would change a couple of things that I missed in my first review, sorry for that (thanks for pointing me to this, Pablo):

- For consistency, you should clamp both the position and the width - probably the latter needs a separate method as it needs to take pos + width into account
- I would give these methods a more descriptive name
- I would add them to `Track` instead of `CaptureViewElement` to make it clear that this is - for now - a change that is somewhat locally to the tracks.

Comment (karupayun) [reply src/OrbitGl/CaptureViewElement.cpp:131]: > Each CaptureViewElement pushes its bounding rect onto the batcher before drawing, and pops it afterwards, that way the draw methods are basically unaffected.

Right! I forgot about Translations. Your idea was to push the size in addition to the position as it's currently working?

> and it might also have significant performance impact.

Why?

> So in conclusion, I think this localized fix is also fine for me. 

I'm not sure if I agree of having a temporal solution just for Tracks. I feel that it will add some mess to the code that might be harder to understand in the future. My first idea of just adding a margin rectangle is more consistent with the current mess while we plan this Translation proper solution.

**Comment by pierricgimmig** [comment]
Thanks for your comments @karupayun and @reichlfl , I'll get back to this this week. Feel free to implement some of the fixes mentioned in the comments yourself btw. Thanks!

File: src/OrbitGl/CallstackThreadBar.cpp (hunk 1 above)
Comment (pierricgimmig) [reply src/OrbitGl/CallstackThreadBar.cpp]: Rebased to use Pablo's fix.

File: src/OrbitGl/CallstackThreadBar.cpp [hunk 3]
```diff
@@ -118,12 +118,15 @@ void CallstackThreadBar::DoUpdatePrimitives(PrimitiveAssembler& primitive_assemb
...
+
   if (!picking) {
     // Draw all callstack samples.
     auto action_on_callstack_events = [&](const CallstackEvent& event) {
       const uint64_t time = event.timestamp_ns();
       ORBIT_CHECK(time >= min_tick && time <= max_tick);
-      const Vec2 pos(GetPixelNumber(event.timestamp_ns(), resolution_in_pixels, min_tick, max_tick),
+      const Vec2 pos(view_element_x + GetPixelNumber(event.timestamp_ns(), resolution_in_pixels,
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/CallstackThreadBar.cpp]: Rebased to use Pablo's fix.

File: src/OrbitGl/CallstackThreadBar.cpp [hunk 4]
```diff
@@ -167,9 +170,10 @@ void CallstackThreadBar::DoUpdatePrimitives(PrimitiveAssembler& primitive_assemb
     auto action_on_callstack_events = [&, this](const CallstackEvent& event) {
       const uint64_t time = event.timestamp_ns();
       ORBIT_CHECK(time >= min_tick && time <= max_tick);
-      const Vec2 pos(
-          GetPixelNumber(time, resolution_in_pixels, min_tick, max_tick) - kPickingBoxOffset,
-          GetPos()[1]);
+      const Vec2 pos(view_element_x +
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/CallstackThreadBar.cpp]: Rebased to use Pablo's fix.

File: src/OrbitGl/TimeGraph.cpp [hunk 5]
```diff
@@ -682,21 +684,21 @@ void TimeGraph::DoUpdateLayout() {
...
+  // | T |     SPACE TRACKS - TIMELINE      H       |
+  // |   |--------------------------------| T |-----|
+  // | M |                                |   |  S  |
+  // | A |                                | M |  L  |
+  // | R |         TRACK CONTAINER        | A |  I  |
+  // | G |                                | R |  D  |
+  // | I |                                | G |  E  |
+  // | N |                                | I |  R  |
+  // |___|________________________________|_N_|_____|
+  // |       HORIZONTAL SLIDER                |
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/TimeGraph.cpp]: done, and also made the horizontal slider respect the margin.

File: src/OrbitGl/include/OrbitGl/StaticTimeGraphLayout.h [hunk 6]
```diff
@@ -134,6 +135,7 @@ class StaticTimeGraphLayout : public TimeGraphLayout {
   constexpr static float kRoundingRadius = 8.f;
   constexpr static float kRoundingNumSides = 16;
   constexpr static float kTextOffset = 5.f;
+  constexpr static float kLeftMargin = 0.f;
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/include/OrbitGl/StaticTimeGraphLayout.h]: There is now a value of 100.f.

File: src/OrbitGl/TimeGraph.cpp [hunk 7]
```diff
@@ -710,8 +712,8 @@ void TimeGraph::UpdateChildrenPosAndContainerSize() {
   float timegraph_current_y = GetPos()[1];
   const float total_right_margin = layout_->GetRightMargin() + vertical_slider_->GetWidth();
 
-  timeline_ui_->SetWidth(GetWidth() - total_right_margin);
-  timeline_ui_->SetPos(timegraph_current_x, timegraph_current_y);
+  timeline_ui_->SetWidth(GetWidth() - total_right_margin - layout_->GetLeftMargin());
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/TimeGraph.cpp:715]: I'll let you implement it if that's ok with you, seems like a big enough change for a new PR at least.

File: src/OrbitGl/include/OrbitGl/CaptureViewElement.h [hunk 8]
```diff
@@ -50,6 +50,8 @@ class CaptureViewElement : public Pickable, public AccessibleInterfaceProvider {
   [[nodiscard]] Vec2 GetSize() const { return Vec2(GetWidth(), GetHeight()); }
   [[nodiscard]] virtual bool ShouldBeRendered() const { return GetVisible(); }
 
+  [[nodiscard]] float HorizontalClamp(float pos_x) const;
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/include/OrbitGl/CaptureViewElement.h:53]: Not sure HorizontalClampIntoView conveys the information that we are in World coordinates either. Maybe ClampToWorldExtentsX() ?

**Comment by pierricgimmig** [comment]
Hi @reichlfl , hi @karupayun ,

I implemented most the suggested changes, and more. The original idea was to make it easier to change the layout, specifically, I'm interested in drawing thing at the left of thread tracks. With this PR, it is now the responsibility of capture view elements to make sure they draw within their allocated space. In my view, this is better than the existing cover-up of the right margin because it forces us to draw strictly what is needed, and it frees up the space on both the left and right margin to draw other things easily, without z-value tricks. Consequently, I removed the right-margin rectangle draw call since it's not required anymore.

There is one functional change in this PR: timeline labels are not partially drawn anymore. If this is a blocker, then we can either re-add the margin cover-up only for the timeline-ui, or come up with a better solution that would allow us to trim strings. Frankly, I really hope it won't be a blocker... 

I changed the `HorizontalClamp` method to `ClampToWorldExtentsX`, but I left it in CaptureViewElement.h because it's dealing with private members of this class, and that any draw element could eventually use it.

This PR also fixes several assumptions/bugs in UI tests.

https://user-images.githubusercontent.com/3687222/205510416-a047dc14-acb1-45ec-bc02-0b83ba6ad972.mp4

File: src/OrbitGl/include/OrbitGl/CaptureViewElement.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/include/OrbitGl/CaptureViewElement.h]: Done.

File: src/OrbitGl/CaptureViewElement.cpp (hunk 2 above)
Comment (pierricgimmig) [reply src/OrbitGl/CaptureViewElement.cpp]: I implemented Florian's suggestions. @karupayun, please see rationale below about removing margin rectangles.

**Review by karupayun — COMMENTED** [review]
Hey Pierric. I really like the idea of not depending of the z-layers + margins to assure containers are not drawing outside of their respective area.

However clamping doesn't work. It works for lines parallel to the axis and rectangles but stop working for everything else. In our case, as far as I know, it only doesn't work for the triangles in the GpuDebugMarker, see screenshots of how they change their shape and the text doesn't fit anymore.

Screen Shot 2022-12-05 at 19 19 04
Screen Shot 2022-12-05 at 19 19 14

Screen Shot 2022-12-05 at 19 19 29

I started a prototype for properly solving it in the CPU:
1- Each container tells the Batcher their pos + size.
2- The Batcher taking into account the position + size of the ancestors:
   - If all points are inside, they send the primitive as it is
   - If not, it intersects the shape with the drawing zone of the container doing segment to segment intersection (there are some details that we can discuss if you are interested)
   - Finally it draws the new shape (that potentially might have more points).
Assuming that most primitives will fit in the container, the performance regression shouldn't be a big deal. 
   
It's a bunch of work that will take a bit of time and I don't have it. We discussed with @reichlfl and we agreed of properly solve the shape intersection issue in the Gpu (that actually makes more sense). Florian will try to do some experiments to see if we have performance regression there, but the idea is setting the position + size of the container into account and let the Gpu do the work for us.

If this idea doesn't work we can work in the Cpu solution. Meanwhile, I would say setting a left-margin and if you need to draw on top of the margin, I would say creating a new CaptureViewElement who draws on top (what do you want to draw there?)

What do you think? Do you have another suggestion?

**Comment by pierricgimmig** [comment]
> What do you think? Do you have another suggestion?

It seems the only breaking part of this PR is for the GpuDebugMarkers. The fix is fairly easy to implement, essentially modifying the clipping code in TimerTrack. Let's say that was done, would this PR get approved?

**Comment by karupayun** [comment]
> > What do you think? Do you have another suggestion?
> 
> It seems the only breaking part of this PR is for the GpuDebugMarkers. The fix is fairly easy to implement, essentially modifying the clipping code in TimerTrack. Let's say that was done, would this PR get approved?

Are you also planning to erase the top margin? Because if so, I don't understand how that can be solved modifying the clipping code in TimerTrack. Some Triangles can have 4 or even 5 vertices after clipping.

I don't find it consistently and we will have bugs in the feature (I guess for example that we are also breaking the white arrow between thread state dependencies).

I would say that if you want that, we should erase also the generic AddLine and have 2 methods AddHorizontalLine and AddVerticalLine, but I don't think that this is the right way, interested in @reichlfl opinion, since tomorrow I start vacation until EOY and I won't be looking for this.

PS: Both the green-overlay and the iterators are being drawn on top of the margins now.

File: src/OrbitGl/SliderTest.cpp [hunk 9]
```diff
@@ -346,10 +376,14 @@ TEST(Slider, MouseMoveRequestRedraw) {
...
 
-  std::shared_ptr<GlHorizontalSlider> slider{
-      std::make_shared<GlHorizontalSlider>(nullptr, viewport, tester.GetLayout(), nullptr)};
+  std::shared_ptr<GlHorizontalSlider> slider =
+      CreateTestSlider<GlHorizontalSlider>(viewport, tester.GetLayout());
   tester.SimulatePreRender(slider.get());
 
+  // In CreateTestSlider, we have called SetWidth() which sets "update_primitives_requested_" to
+  // true. Simulate a draw loop to reinitialize the draw flags before testing their values below.
+  tester.SimulateDrawLoop(slider.get(), true, true);
```
Comment (reichlfl) [inline src/OrbitGl/SliderTest.cpp:385]: I think with this line you don't need the `SimulatePreRender` before.

**Comment by reichlfl** [comment]
I have to admit that I'm not sure about this change. I agree with the intent, and I definitely like the changes that we do not assume all tracks are positioned at 0 and span the whole width. So I would definitely merge these parts. I also think it's a good idea to provide facilities in `CaptureViewElement` to restrict their contents to their own body, and use this where possible.

I don't think that a `CaptureViewElement` should have to ensure this by manually clipping triangles though. This feels like too much effort put onto the implementor of UI elements. It would be doable for GPU tracks, but it already gets complicated if you'd want to consider correct clipping on top/bottom as well. I think it should be responsibility of the framework to ensure drawing of element content is restricted to their body...

For me, the ideal state which would make a lot of calculations obsolete and would allow more flexible layouting would be:
1) Element positions are always relative, so you don't need to care about global positions
2) Elements do their best to restrict drawing to their body, but it is ultimately enforced by the renderer.

I'm experimenting with a solution for 2) by restricting the draw area with the stencil buffer before each element. I'll let you know if this comes at an acceptable performance hit.

There's also other non-breaking, but weird implications to these margins. As Pablo pointed out, the green selection is still visible in the margin, but not behind the cover-up rectangles in the timeline UI...

Meanwhile, I would actually suggest the following:
1) Stick with the cover-up rectangles for the left / right margins for the full height and hope to get rid of it in the future
2) Keep your changes regarding position and clipping - they don't break stuff, and if we can restrict the rendering to the element body on the CPU already, all the better
3) Keep the GPU track as-is, it'll be covered up by the rectangles in the margins.

WDYT?

File: src/OrbitGl/TimeGraph.cpp [hunk 10]
```diff
@@ -883,10 +887,19 @@ void TimeGraph::DrawMarginsBetweenChildren(
   primitive_assembler.AddBox(MakeBox(timeline_margin_pos, timeline_margin_size),
                              GlCanvas::kZValueTimeBar, GlCanvas::kBackgroundColor);
 
-  // Margin between the Tracks and the vertical scrollbar.
-  Vec2 right_margin_pos{GetWidth() - GetRightMargin(), GetPos()[1]};
-  Quad box = MakeBox(right_margin_pos, GetSize() - (right_margin_pos - GetPos()));
-  primitive_assembler.AddBox(box, GlCanvas::kZValueMargin, GlCanvas::kBackgroundColor);
+  // Right margin mask for Timeline.
+  float right_margin_width = GetRightMargin();
+  float right_margin_height = GetHeight() - timeline_ui_->GetHeight();
```
Comment (reichlfl) [inline src/OrbitGl/TimeGraph.cpp:892]: Shouldn't this be `GetHeight()`? The margin spans across the timeline and is positioned at y==0, so why subtract the timeline height?

Comment (reichlfl) [reply src/OrbitGl/TimeGraph.cpp:892]: Looking at the code above - should this be `GetHeight() - horizontal_slider_->GetHeight()`? The margin ends above the slider.

By the way, what's the reasoning behind the different behaviors of left and right margin (left one goes all the way to the bottom, right one stops above the slider)? Wouldn't it be more consistent to have the horizontal slider also stop at the margin?

**Review by reichlfl — APPROVED** [review]
Thanks for the changes! PTAL at the open comments and fix if needed. Otherwise this LGTM, sorry if this was a lot of back and forth - layouting is still more complicated than it needs to be I'm afraid. I'll loop you in on any changes I do to restricting the drawing area to individual elements, maybe we can then remove the margin boxes in the future.

File: src/OrbitGl/TimeGraph.cpp (hunk 10 above)
Comment (pierricgimmig) [reply src/OrbitGl/TimeGraph.cpp:892]: Thanks for the review @reichlfl . The reasoning is that I found it weird to have the horizontal slider not touch the horizontal slider, feel free to modify. I've done enough work on this PR :-)
